File: leetcode/src/com/fqh/滑动窗口/MST1718.java
package com.fqh.滑动窗口;

import java.util.HashMap;

/**
 * @author ikun
 * @version v1.0.0
 * @since 2024/4/25 23:12
 **/
public class MST1718 {

    // https://leetcode.cn/problems/shortest-supersequence-lcci/description/

    public int[] shortestSeq(int[] big, int[] small) {
        int l = -1, r = 100001;
        var map = new HashMap<Integer, Integer>();
        for (int x : small) map.merge(x, 1, Integer::sum);
        int valid = small.length;
        int j = 0;
        for (int i = 0; i < big.length; i++) {
            int x = big[i];
            if (map.getOrDefault(x, 0) > 0) {
                valid -= 1;
            }
            map.merge(x, -1, Integer::sum); // 不在small里面的数字cnt一定 < 0
            if (valid == 0) { // 一个合法的区间
                // 尝试缩小区间，注意只移除不在small里面的数字，也就是map[x] < 0的数字
                while (j < i && map.get(big[j]) < 0) {
                    map.merge(big[j], 1, Integer::sum);
                    j++;
                }
                if (i - j < r - l) {
                    r = i;
                    l = j;
                }
                // j一定是在small里面的数字，将map[big[j]]++ valid++
                map.merge(big[j], 1, Integer::sum);
                j++;
                valid++;
            }
        }
        if (l == -1 || r == 100001) return new int[0];
        return new int[]{l, r};
    }
}

File: py_code/技巧/滑动窗口/MST_17_18.py
from itertools import *
from collections import *
from math import *
from cmath import *
from heapq import *
from functools import *
from bisect import *
from typing import List

MOD = 10 ** 9 + 7
inf = float('inf')
def PF(a):
    return [0] + list(accumulate(a))



# https://leetcode.cn/problems/shortest-supersequence-lcci/description/
class Solution:
    def shortestSeq(self, big: List[int], small: List[int]) -> List[int]:
        ans_left, ans_right = -1, 100001
        cnt_s = Counter(small)
        cnt = len(small)
        j = 0
        for i, x in enumerate(big):
            if cnt_s[x] > 0:
                cnt -= 1
            cnt_s[x] -= 1
            if cnt == 0:
                while j < i and cnt_s[big[j]] < 0:
                    cnt_s[big[j]] += 1
                    j += 1
                if i - j < ans_right - ans_left:
                    ans_left, ans_right = j, i
                cnt_s[big[j]] += 1
                j += 1
                cnt += 1
        if ans_left == -1 or ans_right == 100001:
            return []
        return [ans_left, ans_right]

